### ISSUE TITLE
Implement N50 testing action

### ISSUE THREAD

**Issue by jonperdomo** [WGLab/LongReadSum#9]
Implement an action that will test the N50 calculation.

**Comment by jonperdomo** [comment]
LongReadSum, NanoPlot, and my own MATLAB implementation yield identical results for N50 using sample input FASTA files (individual and combined). These will be used for testing the FASTA output:
| File | N50 |
| --- | --- |
| all_chr.hap1.cns.fa | 23758872 |
 | all_chr.hap2.cns.fa | 21856627 |
| Both files | 23544988 |

**Comment by jonperdomo** [comment]
Results from our 50-read sample dataset
(Mine = My own MATLAB N50 implementation)

| Tool | File | N50 | 
| --- | --- | --- |
| LongReadSum | FASTQ | 8731 |
| NanoPlot | FASTQ | 8733 |
| Mine | FASTQ | 8733 |
| LongReadSum | BAM | 7415 |
| NanoPlot | BAM | 6398 |
| Mine | BAM | 6398 |
| PycoQC | BAM + sequencing_summary.txt | Basecalls = 8730<br />Alignments = 3500 |
| MinIONQC | sequencing_summary.txt| 8733 |

**Comment by kaichop** [comment]
Also check number of alignment (mapped and unmapped), total bases, total
mapped bases.
N50 is a value that is not too informative by itself; it is important in
genome assembly, but not in characterizing sequenced reads.
You can extract fastq/fasta file from BAM file as well using bedtools,
which should help diagnose the discrepancies.

**Comment by jonperdomo** [comment]
Results from our 50-read sample dataset
(Mine = My own MATLAB N50 implementation)

| Tool | File | Read count | Total bases | Total mapped bases | Number of aligned reads | Number of unaligned reads | N50
| --- | --- | --- | --- | --- | --- | --- | --- |
| LongReadSum | FASTQ | 50 | 340189 | NA | NA | NA |  8731
| NanoPlot | FASTQ | 50 | 340189 | NA | NA | NA | 8733 |
| Mine | FASTQ | | | | | | 8733
| LongReadSum | BAM | 50 | 340189 | 304323 | 50 | 0 | 7415
| NanoPlot | BAM | 106<br />(=primary + supplementary alignments?) | 493817 | 314962 | NA | NA | 6398 |
| Mine | BAM | | | | | | 6398
| PycoQC | BAM + sequencing_summary.txt | | | | | | Basecalls = 8730<br />Alignments = 3500
| MinIONQC | sequencing_summary.txt| | | | | | 8733
| LongReadSum | FASTQ from BAM<br />(via bedtools) | | | | |
| NanoPlot | FASTQ from BAM<br />(via bedtools) | | | | |
| Mine | FASTQ from BAM<br />(via bedtools) | | | | |

**Comment by kaichop** [comment]
Are the BAM results for NanoPlot directly reported by this software?
What is the raw output? (does it offer any clues why read count is 106?)

**Comment by jonperdomo** [comment]
No, I run NanoPlot separately to compare results. This is the raw NanoPlot output for the BAM file:

> General summary:         
> Average percent identity:        87.4
> Fraction of bases aligned:         0.6
> Mean read length:             4,658.7
> Mean read quality:                9.7
> Median percent identity:         88.3
> Median read length:           3,579.5
> Median read quality:              9.2
> Number of reads:                106.0
> Read length N50:              6,398.0
> STDEV read length:            3,829.5
> Total bases:                493,817.0
> Total bases aligned:        314,962.0
> Number, percentage and megabases of reads above quality cutoffs
> >Q5:	106 (100.0%) 0.5Mb
> >Q7:	99 (93.4%) 0.5Mb
> >Q10:	32 (30.2%) 0.2Mb
> >Q12:	15 (14.2%) 0.1Mb
> >Q15:	5 (4.7%) 0.0Mb
> Top 5 highest mean basecall quality scores and their read lengths
> 1:	17.9 (1239; bf3ca0c2-c10a-40f5-bf35-4b6d2ba8479a)
> 2:	17.4 (3242; ff7bd978-ce99-42fb-a332-9fdf78f90701)
> 3:	16.9 (3103; e7a151f2-7620-4d0e-9d84-41e8dc4cab5f)
> 4:	16.5 (6557; e7a151f2-7620-4d0e-9d84-41e8dc4cab5f)
> 5:	16.4 (7050; ff7bd978-ce99-42fb-a332-9fdf78f90701)
> Top 5 longest reads and their mean basecall quality score
> 1:	19696 (7.5; a3e904d3-7f4d-4ba8-bb61-77505f026a1f)
> 2:	17825 (11.1; cf8eab27-821c-44bd-bfac-023299c37aed)
> 3:	16511 (9.7; 8631ced5-cfa0-4ff1-af7b-5abba3f96fa9)
> 4:	15263 (9.2; 2960e123-09ed-4596-ae15-10027f4b375f)
> 5:	14473 (8.5; ecee2aa6-765e-4ba1-a67f-7130f6e87517)


LongReadSum has more information on read counts for alignments, which indicate that possibly the supplementary aligments are being included in the total read count:
![image](https://user-images.githubusercontent.com/14855676/174495103-32a8053f-05aa-490e-8878-6fd70fd36e88.png)

**Comment by jonperdomo** [comment]
I tested this by filtering out unmapped, secondary, and supplementary alignments via the SAMTools view function. When filtered out, the number of reads output by NanoPlot is correct (=50):

> General summary:         
> Average percent identity:        91.7
> Fraction of bases aligned:         0.5
> Mean read length:             6,803.8
> Mean read quality:               10.1
> Median percent identity:         93.2
> Median read length:           5,194.0
> Median read quality:              9.6
> Number of reads:                 50.0
> Read length N50:              8,733.0
> STDEV read length:            4,256.2
> Total bases:                340,189.0
> Total bases aligned:        161,334.0
> Number, percentage and megabases of reads above quality cutoffs
> >Q5:	50 (100.0%) 0.3Mb
> >Q7:	50 (100.0%) 0.3Mb
> >Q10:	18 (36.0%) 0.1Mb
> >Q12:	7 (14.0%) 0.0Mb
> >Q15:	2 (4.0%) 0.0Mb
> Top 5 highest mean basecall quality scores and their read lengths
> 1:	16.5 (6557; e7a151f2-7620-4d0e-9d84-41e8dc4cab5f)
> 2:	16.4 (7050; ff7bd978-ce99-42fb-a332-9fdf78f90701)
> 3:	14.3 (6925; e621bda1-0721-402e-9ca4-49cd42ff41cc)
> 4:	13.9 (5314; bad54463-4c8b-4007-8375-2df0687508f8)
> 5:	13.6 (3505; 00a29863-c29f-461c-bd56-1943fd2711a8)
> Top 5 longest reads and their mean basecall quality score
> 1:	19696 (7.5; a3e904d3-7f4d-4ba8-bb61-77505f026a1f)
> 2:	17825 (11.1; cf8eab27-821c-44bd-bfac-023299c37aed)
> 3:	16511 (9.7; 8631ced5-cfa0-4ff1-af7b-5abba3f96fa9)
> 4:	15263 (9.2; 2960e123-09ed-4596-ae15-10027f4b375f)
> 5:	14473 (8.5; ecee2aa6-765e-4ba1-a67f-7130f6e87517)

I used LongReadSum to verify that this filtered BAM file only contains the primary alignments:
![image](https://user-images.githubusercontent.com/14855676/174495760-cbaa098e-0b18-4cbb-ad52-17cbc66b46f9.png)

**Comment by kaichop** [comment]
okay I see, thank you.

**Comment by liuqianhn** [comment]
@jonperdomo 
"I used LongReadSum to verify that this filtered BAM file only contains the primary alignments:"
LongReadSum has a filter where you can set other alignments to be false and only the alignments with true will be considered. You can change the program to allow users to specify those filter setting so that you can focus on the primary alignments only.
